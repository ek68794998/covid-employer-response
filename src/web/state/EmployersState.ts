import { EmployerRecord } from "../../common/EmployerRecord";

export interface EmployersState {
	items: EmployerRecord[];
}