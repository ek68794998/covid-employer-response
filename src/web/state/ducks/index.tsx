import { combineReducers } from "redux";

import { getLocalizedStrings } from "./localization/reducers";

export default combineReducers({
	strings: getLocalizedStrings,
});