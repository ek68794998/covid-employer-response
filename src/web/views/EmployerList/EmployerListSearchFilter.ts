import { EmployerRecord } from "../../../common/EmployerRecord";

export class EmployerListSearchFilter {
	public text: string = "";

	public static isMatch(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		return e.name.toLowerCase().indexOf(f.text.toLowerCase()) >= 0;
	}
}