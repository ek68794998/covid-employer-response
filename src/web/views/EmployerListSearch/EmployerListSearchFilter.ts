import { EmployerRecord } from "../../../common/EmployerRecord";

export class EmployerListSearchFilter {
	public static readonly SMALL_BOUNDARY: number = 1000;

	public static readonly LARGE_BOUNDARY: number = 100000;

	public international: boolean = true;

	public large: boolean = true;

	public medium: boolean = true;

	public national: boolean = true;

	public small: boolean = true;

	public text: string = "";

	public static isMatch(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!this.isMatchForLocation(f, e)) {
			return false;
		}

		if (!this.isMatchForEmployeeCount(f, e)) {
			return false;
		}

		return this.isMatchForFullText(f, e);
	}

	private static isMatchForEmployeeCount(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!e.employeesBefore) {
			return false;
		}

		const getNumberIsBetween = (v: number, lower: number, upper: number): boolean =>
			v >= lower && v <= upper;

		const getNumberIsMatch = (v: number): boolean => {
			if (!f.small && v <= this.SMALL_BOUNDARY) {
				return false;
			}

			if (!f.large && v >= this.LARGE_BOUNDARY) {
				return false;
			}

			if (!f.medium
				&& getNumberIsBetween(v, this.SMALL_BOUNDARY, this.LARGE_BOUNDARY)) {

				return false;
			}

			return true;
		};

		const lowerBound: number = e.employeesBefore.lowerBound || 0;
		const upperBound: number = Math.max(e.employeesBefore.upperBound, lowerBound);

		if (e.employeesBefore.type === "approximately"
			|| e.employeesBefore.type === "exactly"
			|| lowerBound <= 0) {

			return getNumberIsMatch(upperBound);
		}

		if (upperBound <= 0) {
			return getNumberIsMatch(lowerBound);
		}

		const rangeIncludesSmall: boolean = lowerBound < this.SMALL_BOUNDARY;

		const rangeIncludesMedium: boolean =
			getNumberIsBetween(lowerBound, this.SMALL_BOUNDARY, this.LARGE_BOUNDARY)
			|| getNumberIsBetween(upperBound, this.SMALL_BOUNDARY, this.LARGE_BOUNDARY);

		const rangeIncludesLarge: boolean = upperBound > this.LARGE_BOUNDARY;

		return (f.small && rangeIncludesSmall)
			|| (f.medium && rangeIncludesMedium)
			|| (f.large && rangeIncludesLarge);
	}

	private static isMatchForFullText(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!f.text) {
			return true;
		}

		const fieldsToSearch: (string | undefined)[] = [
			e.name.toLowerCase(),
			e.location?.city.toLowerCase(),
			e.shortName?.toLowerCase(),
			e.ticker?.toLowerCase(),
		];

		if (e.aliases) {
			e.aliases.forEach((a: string) => fieldsToSearch.push(a));
		}

		return fieldsToSearch.some((field?: string) => field && field.indexOf(f.text.toLowerCase()) >= 0);
	}

	private static isMatchForLocation(f: EmployerListSearchFilter, e: EmployerRecord): boolean {
		if (!e.location) {
			return false;
		}

		if ((!f.international && e.location.international) || (!f.national && !e.location.international)) {
			return false;
		}

		return true;
	}
}