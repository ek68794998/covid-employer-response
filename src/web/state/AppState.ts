import { EmployersState } from "./EmployersState";

import { LocalizedStrings } from "../../common/LocalizedStrings";

export interface AppState {
	employers: EmployersState;

	environment: string;

	strings: LocalizedStrings;
}