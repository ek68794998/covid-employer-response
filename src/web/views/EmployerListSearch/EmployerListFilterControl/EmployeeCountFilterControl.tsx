import React from "react";
import { useSelector } from "react-redux";

import { LocalizedStrings } from "../../../../common/LocalizedStrings";

import { getStrings } from "../../../state/ducks/localization/selectors";

import { EmployerListSearchFilter } from "../EmployerListSearchFilter";
import { EmployerListSearchFilterProps } from "../EmployerListSearchFilterProps";

import EmployerListFilterControl, { ChildSelectorProps } from "./EmployerListFilterControl";

import "./EmployerListFilterControl.scss";

const EmployeeCountFilterControl: React.FC<EmployerListSearchFilterProps> =
	(props: EmployerListSearchFilterProps): React.ReactElement => {
		const strings: LocalizedStrings = useSelector(getStrings);

		const { filter, onUpdateFilter } = props;

		const getDisplayText = (): string => {
			let newDisplayText: string = "";

			if (!(filter.small && filter.medium && filter.large)) {
				const displayValues: string[] = [];

				if (filter.small) {
					displayValues.push(strings.filters.employeesSmall);
				}

				if (filter.medium) {
					displayValues.push(strings.filters.employeesMedium);
				}

				if (filter.large) {
					displayValues.push(strings.filters.employeesLarge);
				}

				newDisplayText = displayValues.join(", ");
			}

			return newDisplayText || strings.filters.employeesDefault;
		};

		const onClear = (): void => {
			filter.small = true;
			filter.medium = true;
			filter.large = true;

			onUpdateFilter();
		};

		const selectorPropsList: ChildSelectorProps[] = [
			{
				initialValue: filter.small,
				label: strings.filters.employeesSmall,
				setValue: (value: boolean): void => onUpdateFilterValue({ small: value }),
			},
			{
				initialValue: filter.medium,
				label: strings.filters.employeesMedium,
				setValue: (value: boolean): void => onUpdateFilterValue({ medium: value }),
			},
			{
				initialValue: filter.large,
				label: strings.filters.employeesLarge,
				setValue: (value: boolean): void => onUpdateFilterValue({ large: value }),
			},
		];

		const onUpdateFilterValue = (updates: Partial<EmployerListSearchFilter>): void => {
			Object.assign(filter, updates);
			onUpdateFilter();
		};

		return (
			<EmployerListFilterControl
				getDisplayText={getDisplayText}
				isActive={!(filter.small && filter.medium && filter.large)}
				onClear={onClear}
				popupTitle={strings.filters.employeesDefault}
				selectorPropsList={selectorPropsList}
			/>
		);
	};

export default EmployeeCountFilterControl;
