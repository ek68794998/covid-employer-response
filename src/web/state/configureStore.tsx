import { AnyAction, applyMiddleware, compose, createStore, Store } from "redux";
import thunk from "redux-thunk";

import { Production } from "../../common/constants/EnvironmentConstants";

import { AppState } from "./AppState";
import rootReducer from "./ducks";

const configureStore = (preloadedState: Partial<AppState>): Store<AppState, AnyAction> => {
	const isProd: boolean = process.env.NODE_ENV === Production;
	const composeEnhancers: any =
		(!isProd && typeof window !== "undefined" && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__) || compose;

	const store: Store<AppState, AnyAction> = createStore(
		rootReducer,
		preloadedState,
		composeEnhancers(applyMiddleware(thunk)),
	);

	if (module.hot) {
		// Enable Webpack hot module replacement for reducers
		module.hot.accept("./ducks", () => {
			const nextRootReducer: any = require("./ducks").default;
			store.replaceReducer(nextRootReducer);
		});
	}

	return store;
};

export default configureStore;