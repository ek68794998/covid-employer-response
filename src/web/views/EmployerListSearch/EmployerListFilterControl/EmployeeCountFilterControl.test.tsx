import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer, { ReactTestRendererJSON } from "react-test-renderer";
import { AnyAction, Store } from "redux";

import { AppState } from "../../../state/AppState";
import configureStore from "../../../state/configureStore";

import { EmployerListSearchFilter } from "../EmployerListSearchFilter";

import EmployeeCountFilterControl from "./EmployeeCountFilterControl";

describe("<EmployeeCountFilterControl />", () => {
	test("renders without exploding", () => {
		const store: Store<AppState, AnyAction> = configureStore({
			strings: {
				filters: {
					employeesDefault: "è₥ƥℓô¥èèƨÐèƒáúℓƭ",
					employeesLarge: "è₥ƥℓô¥èèƨ£ářϱè",
					employeesMedium: "è₥ƥℓô¥èèƨMèδïú₥",
					employeesSmall: "è₥ƥℓô¥èèƨ§₥áℓℓ",
				},
			},
		});

		const renderedValue: ReactTestRendererJSON | null =
			renderer.create(
				<Provider store={store}>
					<BrowserRouter>
						<EmployeeCountFilterControl
							filter={new EmployerListSearchFilter()}
							onUpdateFilter={(): void => { /* Do nothing. */ }}
						/>
					</BrowserRouter>
				</Provider>,
			).toJSON();

		expect(renderedValue).toMatchSnapshot();
	});
});
