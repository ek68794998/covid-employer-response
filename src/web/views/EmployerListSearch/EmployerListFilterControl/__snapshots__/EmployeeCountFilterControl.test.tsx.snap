// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EmployeeCountFilterControl /> renders without exploding 1`] = `
<div
  className="EmployerListFilterControl__Container"
>
  <button
    className="EmployerListFilterControl__Button"
    onClick={[Function]}
  >
    <div>
      è₥ƥℓô¥èèƨÐèƒáúℓƭ
    </div>
  </button>
</div>
`;
